enterprise/server/tasksize/tasksize.go

```diff
@@ -64,20 +93,141 @@ func Register(env environment.Env) error {
-	// TODO(bduffany): Implement
-	return status.UnimplementedError("not implemented")
+	if !*adaptiveSizingEnabled {
+		return nil
+	}
+	// If we are missing CPU/memory stats, do nothing. This is expected for
+	// tasks that either completed too quickly to get a sample of their CPU/mem
+	// usage, or tasks where the workload isolation type doesn't yet support the
+	// Stats() API.
+	stats := summary.GetUsageStats()
+	if stats.GetCpuNanos() == 0 || stats.GetPeakMemoryBytes() == 0 {
+		return nil
+	}
+	md := summary.GetExecutedActionMetadata()
+	execDuration := md.GetExecutionCompletedTimestamp().AsTime().Sub(md.GetExecutionStartTimestamp().AsTime())
+	// If execution duration is missing or invalid, we won't be able to compute
+	// milli-CPU usage.
+	if execDuration <= 0 {
+		return status.InvalidArgumentErrorf("execution duration is missing or invalid")
+	}
+	// Compute milliCPU as CPU-milliseconds used per second of execution time.
+	milliCPU := int64((float64(stats.GetCpuNanos()) / 1e6) / execDuration.Seconds())
+	key, err := s.taskSizeKey(ctx, cmd)
+	if err != nil {
+		return err
+	}
+	pipe := s.rdb.TxPipeline()
+	pipe.HSet(ctx, key, redisMemHashKey, stats.GetPeakMemoryBytes())
+	pipe.HSet(ctx, key, redisCPUHashKey, milliCPU)
+	pipe.Expire(ctx, key, redisKeyExpiration)
+	_, err = pipe.Exec(ctx)
+	return err
+}
+
+func (s *taskSizer) lastRecordedSize(ctx context.Context, task *repb.ExecutionTask) (*scpb.TaskSize, error) {
+	key, err := s.taskSizeKey(ctx, task.GetCommand())
+	if err != nil {
+		return nil, err
+	}
+	m, err := s.rdb.HGetAll(ctx, key).Result()
+	if err != nil {
+		return nil, err
+	}
+	// Note: if the key is not found, HGETALL returns an empty map, not a
+	// redis.Nil error.
+	if len(m) == 0 {
+		return nil, nil
+	}
+	mInt64, err := redisutil.Int64Values(m)
+	if err != nil {
+		return nil, err
+	}
+	if mInt64[redisMemHashKey] == 0 || mInt64[redisCPUHashKey] == 0 {
+		return nil, status.InternalError("found invalid task size stored in Redis")
+	}
+	return &scpb.TaskSize{
+		EstimatedMemoryBytes: mInt64[redisMemHashKey],
+		EstimatedMilliCpu:    mInt64[redisCPUHashKey],
+	}, nil
+}
+
+func (s *taskSizer) taskSizeKey(ctx context.Context, cmd *repb.Command) (string, error) {
+	// Get group ID (task sizing is segmented by group)
+	u, err := perms.AuthenticatedUser(ctx, s.env)
+	if err != nil {
+		if !perms.IsAnonymousUserError(err) || !s.env.GetAuthenticator().AnonymousUsageEnabled() {
+			return "", err
+		}
+	}
+	groupKey := "ANON"
+	if u != nil {
+		groupKey = u.GetGroupID()
+	}
+	// For now, associate stats with the exact command, including the full
+	// command line, env vars, and platform.
+	// Note: This doesn't account for platform overrides for now
+	// (--remote_header=x-buildbuddy-platform.NAME=VALUE).
+	cmdKey, err := commandKey(cmd)
+	if err != nil {
+		return "", err
+	}
+	return fmt.Sprintf("%s/%s/%s", redisKeyPrefix, groupKey, cmdKey), nil
```

Comment:
So this will look like:

"taskSize/GR1231231241412//usr/bin/gccSHA12312312314" ?